lib/manager/pre-commit/extract.ts

```diff
@@ -0,0 +1,136 @@
+import is from '@sindresorhus/is';
+import yaml from 'js-yaml';
+import { id as githubTagsId } from '../../datasource/github-tags';
+import { id as gitlabTagsId } from '../../datasource/gitlab-tags';
+import { logger } from '../../logger';
+import { SkipReason } from '../../types';
+import { regEx } from '../../util/regex';
+import { PackageDependency, PackageFile } from '../common';
+
+import {
+  matchesPrecommitConfigHeuristic,
+  matchesPrecommitDependencyHeuristic,
+} from './parsing';
+import { PreCommitConfig } from './types';
+
+function determineDatasource(
+  repository: string,
+  domain: string
+): { datasource?: string; registryUrls?: string[] } {
+  if (domain === 'github.com') {
+    logger.debug({ repository, domain }, 'Found github dependency');
+    return { datasource: githubTagsId };
+  }
+  if (domain === 'gitlab.com') {
+    logger.debug({ repository, domain }, 'Found gitlab dependency');
+    return { datasource: gitlabTagsId };
+  }
+  logger.debug(
+    { repository, domain },
+    'Not github.com or gitlab.com, assuming private gitlab-ee.'
+  );
+  return { datasource: gitlabTagsId, registryUrls: [domain] };
```

Comment:
I'm not sure about this assumption. 

What if you called [hostRules.find()](https://github.com/renovatebot/renovate/blob/1c2e64980ce4835da0f7eb2a02154c5ce2e4c5e7/lib/util/host-rules.ts#L90) supplying the hostname and `hostType: 'gitlab'`. If you get a result, you can assume it's a gitlab host. You can do the same for `github` too.

Are you also sure that `registryUrls: [domain]` works? e.g. if the gitlab server is on `gitlab.company.com` then setting `registryUrls: ['company.com']` isn't going to work.

Reply:
if the gitlab server is on `gitlab.company.com` ,the variable `domain` will hold the value `gitlab.company.com`. I only strip the `https` or `git@` from the url, until the first `/` or `:`.

Reply:
I didn't know about `hostRules.find()` but seems a good way to distinguish custom github from gitlab repo's, and detect unlisted hosts.
I'll take a look at how it's been used in pre-commit in other places, and submit the improvement

Reply:
> if the gitlab server is on `gitlab.company.com` ,the variable `domain` will hold the value `gitlab.company.com`. I only strip the `https` or `git@` from the url, until the first `/` or `:`.

If that's the case then we need to change either (a) the variable name, or (b) the variable contents because it's currently misleading/confusing. It's ok to then use "hostName" and make sure the checks are equality ones and not endsWidth (so we don't trigger security errors).

Reply:
@rarkins  , how about flipping it around: query `hostRules.find()` with the `hostName` as query `url`, and then use the returned `hostType` (or nothing) to determine which datasource to use. If nothing returned, skip the dependency.

Reply:
`hostType` is not returned: https://github.com/renovatebot/renovate/blob/1c2e64980ce4835da0f7eb2a02154c5ce2e4c5e7/lib/util/host-rules.ts#L139

A host could have multiple rules and hostTypes, which is part of the reason why.

Reply:
We *could* expand the hostRules API with a new function where you input a hostname and return a list of `hostTypes`, if you think my first suggestion is not elegant

Reply:
ah, didn't see that.
But checking if there is 1 certain hostType, doesn't ensure the hostName of the dependency belongs to that hostType.
I had implemented it already, assuming hostType would be returned:

Reply:
k, how about checking both `url` & `hostType`? Does require multiple `hostRules.find` calls unfortunately

Reply:
@rarkins just pushed an alternative version.
I first test if hostName occurs at least once.
Then I test for which hostType it occurs.